pkg/objstore/swift/swift.go

```diff
@@ -82,215 +190,130 @@ func (c *Container) Name() string {
-
 	return objstore.ObjectAttributes{
-		Size:         headers.ContentLength,
-		LastModified: headers.LastModified,
+		Size:         info.Bytes,
+		LastModified: info.LastModified,
 	}, nil
 }
 
 // Exists checks if the given object exists.
-func (c *Container) Exists(ctx context.Context, name string) (bool, error) {
-	err := objects.Get(c.client, c.name, name, nil).Err
-	if err == nil {
-		return true, nil
-	}
-
-	if _, ok := err.(gophercloud.ErrDefault404); ok {
-		return false, nil
+func (c *Container) Exists(_ context.Context, name string) (bool, error) {
+	found := true
+	_, _, err := c.connection.Object(c.name, name)
+	if c.IsObjNotFoundErr(err) {
+		err = nil
+		found = false
 	}
-
-	return false, err
+	return found, err
 }
 
 // IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
 func (c *Container) IsObjNotFoundErr(err error) bool {
-	_, ok := err.(gophercloud.ErrDefault404)
-	return ok
+	return errors.Is(err, swift.ObjectNotFound)
 }
 
 // Upload writes the contents of the reader as an object into the container.
-func (c *Container) Upload(ctx context.Context, name string, r io.Reader) error {
-	options := &objects.CreateOpts{Content: r}
-	res := objects.Create(c.client, c.name, name, options)
-	return res.Err
-}
-
-// Delete removes the object with the given name.
-func (c *Container) Delete(ctx context.Context, name string) error {
-	return objects.Delete(c.client, c.name, name, nil).Err
-}
-
-func (*Container) Close() error {
-	// Nothing to close.
-	return nil
-}
-
-func parseConfig(conf []byte) (*SwiftConfig, error) {
-	var sc SwiftConfig
-	err := yaml.UnmarshalStrict(conf, &sc)
-	return &sc, err
-}
-
-func authOptsFromConfig(sc *SwiftConfig) gophercloud.AuthOptions {
-	authOpts := gophercloud.AuthOptions{
-		IdentityEndpoint: sc.AuthUrl,
-		Username:         sc.Username,
-		UserID:           sc.UserId,
-		Password:         sc.Password,
-		DomainID:         sc.DomainId,
-		DomainName:       sc.DomainName,
-		TenantID:         sc.ProjectID,
-		TenantName:       sc.ProjectName,
-
-		// Allow Gophercloud to re-authenticate automatically.
-		AllowReauth: true,
+func (c *Container) Upload(_ context.Context, name string, r io.Reader) error {
+	size, err := objstore.TryToGetSize(r)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "could not guess file size, using large object to avoid issues if the file is larger than limit", "name", name, "err", err)
+		// Anything higher or equal to chunk size so the SLO is used.
+		size = c.chunkSize
 	}
-
-	// Support for cross-domain scoping (user in different domain than project).
-	// If a userDomainName or userDomainID is given, the user is scoped to this domain.
-	switch {
-	case sc.UserDomainName != "":
-		authOpts.DomainName = sc.UserDomainName
-	case sc.UserDomainID != "":
-		authOpts.DomainID = sc.UserDomainID
+	var file io.WriteCloser
+	if size >= c.chunkSize {
+		file, err = c.connection.StaticLargeObjectCreateFile(&swift.LargeObjectOpts{
```

Comment:
Could we optionally do this with `DynamicLargeObjectCreateFile` too (with some switch and SLO by default)? I know that SLO is used here for consistency, but e.g. our Swift do not support SLO but support DLO.
Much appreciated with effort BTW!

Reply:
Hi, I suppose it won't hurt to add the support for DLO, but I'll add warning to the docs about it's usage and consequences to scare off the users :)

Reply:
The DLO support is added here https://github.com/thanos-io/thanos/commit/7a51f585a7184cc9a0c5a73ffcd77b23896da026, check it out.

Reply:
The tests passes even with the DLO as I tested it now so hopefully it will work out for you :)